ethers-core/src/types/transaction/eip2718.rs

```diff
@@ -398,6 +401,143 @@ impl From<&Transaction> for TypedTransaction {
     }
 }
 
+impl TypedTransaction {
+    #[allow(dead_code)]
+    fn as_legacy_ref(&self) -> Result<&TransactionRequest, TypedTransactionError> {
+        match self {
+            Legacy(tx) => Ok(tx),
+            _ => Err(TypedTransactionError::WrongTransactionType),
+        }
+    }
+    #[allow(dead_code)]
+    fn as_eip2930_ref(&self) -> Result<&Eip2930TransactionRequest, TypedTransactionError> {
+        match self {
+            Eip2930(tx) => Ok(tx),
+            _ => Err(TypedTransactionError::WrongTransactionType),
+        }
+    }
+    #[allow(dead_code)]
+    fn as_eip1559_ref(&self) -> Result<&Eip1559TransactionRequest, TypedTransactionError> {
+        match self {
+            Eip1559(tx) => Ok(tx),
+            _ => Err(TypedTransactionError::WrongTransactionType),
+        }
+    }
+}
+
+impl TypedTransaction {
+    fn as_eip1559(&self) -> Eip1559TransactionRequest {
+        match self {
+            Eip1559(tx) => tx.clone(),
+            _ => Eip1559TransactionRequest {
+                from: self.from().copied(),
+                to: self.to().cloned(),
+                nonce: self.nonce().copied(),
+                value: self.value().copied(),
+                gas: self.gas().copied(),
+                chain_id: self.chain_id(),
+                data: self.data().cloned(),
+                access_list: self.access_list().cloned().unwrap_or_default(),
+                ..Default::default()
+            },
+        }
+    }
+    fn into_eip1559(self) -> Eip1559TransactionRequest {
+        match self {
+            Eip1559(tx) => tx,
+            _ => (&self).as_eip1559(),
+        }
+    }
+}
+
+impl From<&TypedTransaction> for Eip1559TransactionRequest {
+    fn from(src: &TypedTransaction) -> Eip1559TransactionRequest {
+        src.as_eip1559()
+    }
+}
+
+impl From<TypedTransaction> for Eip1559TransactionRequest {
+    fn from(src: TypedTransaction) -> Eip1559TransactionRequest {
+        src.into_eip1559()
+    }
+}
+
+impl TypedTransaction {
+    fn as_legacy(&self) -> TransactionRequest {
```

Comment:
since we basically clone every field, I think we should take ownership here so the caller can decide if a clone is necessary, otherwise, converting always requires a clone

Reply:
yeah seems like we should not have `as_*` methods then, let's remove them and replace with `into_*`